homeassistant/components/weather/metoffice.py

```diff
@@ -0,0 +1,117 @@
+"""
+Support for UK Met Office weather service.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.metoffice/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.weather import WeatherEntity, PLATFORM_SCHEMA
+from homeassistant.const import CONF_NAME, TEMP_CELSIUS
+from homeassistant.helpers import config_validation as cv
+# Reuse data and API logic from the sensor implementation
+from homeassistant.components.sensor.metoffice import \
+    MetOfficeCurrentData, CONF_MO_API_KEY, CONF_ATTRIBUTION
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['datapoint==0.4.3']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_MO_API_KEY): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Met Office weather platform."""
+    import datapoint as dp
+    datapoint = dp.connection(api_key=config.get(CONF_MO_API_KEY))
+
+    if None in (hass.config.latitude, hass.config.longitude):
+        _LOGGER.error("Latitude or longitude not set in Home Assistant config")
+        return False
+
+    try:
+        site = datapoint.get_nearest_site(longitude=hass.config.longitude,
+                                          latitude=hass.config.latitude)
+    except dp.exceptions.APIException as err:
+        _LOGGER.error("Received error from Met Office Datapoint: %s", err)
+        return False
+
+    if not site:
+        _LOGGER.error("Unable to get nearest Met Office forecast site")
+        return False
+    else:
+        # Get data
+        data = MetOfficeCurrentData(hass, datapoint, site)
+        try:
+            data.update()
+        except (ValueError, dp.exceptions.APIException) as err:
+            _LOGGER.error("Received error from Met Office Datapoint: %s", err)
+            return False
+        add_devices([MetOfficeWeather(site, data, config.get(CONF_NAME))],
+                    True)
+        return True
+
+
+class MetOfficeWeather(WeatherEntity):
+    """Implementation of a Met Office weather condition."""
+
+    def __init__(self, site, data, config):
+        """Initialise the platform with a data instance and site."""
+        self.data = data
+        self.site = site
+
+    def update(self):
+        """Update current conditions."""
+        self.data.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return 'Met Office ({})'.format(self.site.name)
+
+    @property
+    def condition(self):
+        """Return the current condition."""
+        return self.data.data.weather.text
```

Comment:
The `condition` should be mapped to an entry from `CONDITION_CLASSES`. This will make it possible in the future to show a visual representation of the condition in the frontend.